hal/arm/rtt.c

```diff
@@ -0,0 +1,193 @@
+
+#ifndef RTT_CB_SIZE
+#define RTT_CB_SIZE 256
+#endif
+
+
+struct rtt_pipe {
+	const char *name;
+	volatile unsigned char *ptr;
+	unsigned int sz;
+	volatile unsigned int wr;
+	volatile unsigned int rd;
+	unsigned int flags;
+};
+
+
+struct rtt_desc {
+	char tag[16];
+	unsigned int txChannels;
+	unsigned int rxChannels;
+	struct rtt_pipe channels[];
+};
+
+
+static struct {
+	volatile struct rtt_desc *rtt;
+} common;
+
+
+int rtt_check(int chan, rtt_dir_t dir)
+{
+	if (chan < 0) {
+		return -EINVAL;
+	}
+
+	if (common.rtt == NULL) {
+		return -ENODEV;
+	}
+
+	if ((dir == rtt_dir_up) && (chan >= common.rtt->txChannels)) {
+		return -ENODEV;
+	}
+
+	if ((dir == rtt_dir_down) && (chan >= common.rtt->rxChannels)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+
+int _hal_rttRead(int chan, void *buf, unsigned int count)
+{
+	if (rtt_check(chan, rtt_dir_down) < 0) {
+		return -ENODEV;
+	}
+
+	volatile struct rtt_desc *rtt = common.rtt;
+
+	hal_cpuDataMemoryBarrier();
+	chan = rtt->txChannels + chan;
+	volatile unsigned char *srcBuf = rtt->channels[chan].ptr;
+	unsigned char *dstBuf = (unsigned char *)buf;
+	unsigned int sz = rtt->channels[chan].sz - 1;
+	unsigned int rd = rtt->channels[chan].rd & sz;
+	unsigned int wr = rtt->channels[chan].wr & sz;
+	unsigned int todo = count;
+
+	while ((todo != 0) && (rd != wr)) {
+		*dstBuf++ = srcBuf[rd];
+		rd = (rd + 1) & sz;
+		todo--;
+	}
+
+	hal_cpuDataMemoryBarrier();
+	rtt->channels[chan].rd = rd;
+
+	return count - todo;
+}
+
+
+int _hal_rttWrite(int chan, const void *buf, unsigned int count)
+{
+	if (rtt_check(chan, rtt_dir_up) < 0) {
+		return -ENODEV;
+	}
+
+	volatile struct rtt_desc *rtt = common.rtt;
+
+	hal_cpuDataMemoryBarrier();
+	const unsigned char *srcBuf = (const unsigned char *)buf;
+	volatile unsigned char *dstBuf = rtt->channels[chan].ptr;
+	unsigned int sz = rtt->channels[chan].sz - 1;
+	unsigned int rd = (rtt->channels[chan].rd + sz) & sz;
+	unsigned int wr = rtt->channels[chan].wr & sz;
+	unsigned int todo = count;
+
+	/* TODO: Support all buffer modes (currently only trim is used, regardless of flags) */
+	while ((todo != 0) && (rd != wr)) {
+		dstBuf[wr] = *srcBuf++;
+		wr = (wr + 1) & sz;
+		todo--;
+	}
+
+	hal_cpuDataMemoryBarrier();
+	rtt->channels[chan].wr = wr;
+
+	return count - todo;
+}
+
+
+int _hal_rttTxAvail(int chan)
+{
+	if (rtt_check(chan, rtt_dir_up) < 0) {
+		return -ENODEV;
+	}
+
+	volatile struct rtt_desc *rtt = common.rtt;
+
+	hal_cpuDataMemoryBarrier();
+	unsigned int sz = rtt->channels[chan].sz - 1;
+	unsigned int rd = (rtt->channels[chan].rd + sz) & sz;
+	unsigned int wr = rtt->channels[chan].wr & sz;
+
+	if (wr > rd) {
+		return sz + 1 - (wr - rd);
+	}
+	else {
+		return rd - wr;
+	}
+}
+
+
+void _hal_rttReset(int chan, rtt_dir_t dir)
+{
+	if (rtt_check(chan, dir) < 0) {
+		return -ENODEV;
+	}
+
+	volatile struct rtt_desc *rtt = common.rtt;
+
+	hal_cpuDataMemoryBarrier();
+	if (dir == rtt_dir_up) {
+		common.rtt->channels[chan].wr = common.rtt->channels[chan].rd;
```

Comment:
You can use pointer `rtt` instead of `common.rtt`.

Reply:
Done